File: leetcode/0067. Add Binary.py
'''
Given two binary strings, return their sum (also a binary string).

The input strings are both non-empty and contains only characters 1 or 0.

Example 1:

Input: a = "11", b = "1"
Output: "100"
Example 2:

Input: a = "1010", b = "1011"
Output: "10101"
'''
class Solution(object):
    def addBinary(self, a, b):
        """
        :type a: str
        :type b: str
        :rtype: str
        """
        lista=[]
        listb=[]
        for ss in a[::-1]:
            lista.append(ss)
        for ss in b[::-1]:
            listb.append(ss)
        if len(lista)>len(listb):
            lista,listb=lista,listb
        else:
            lista,listb=listb,lista
        i=0
        flag='0'
        while i<len(listb):
            if lista[i]=='1' and listb[i]=='1' and flag=='0':
                lista[i]='0'
                flag='1'
            elif lista[i]=='1' and listb[i]=='1' and flag=='1':
                lista[i]='1'
                flag='1'
            elif lista[i]=='0' and listb[i]=='1' and flag=='1':
                lista[i]='0'
                flag='1'
            elif lista[i]=='1' and listb[i]=='0' and flag=='1':
                lista[i]='0'
                flag='1'
            elif lista[i]=='0' and listb[i]=='1' and flag=='0':
                lista[i]='1'
                flag='0'
            elif lista[i]=='1' and listb[i]=='0' and flag=='0':
                lista[i]='1'
                flag='0'
            elif lista[i]=='0' and listb[i]=='0' and flag=='1':
                lista[i]='1'
                flag='0'
            elif lista[i]=='0' and listb[i]=='0' and flag=='0':
                lista[i]='0'
                flag='0'
            i+=1
        
        while i<len(lista):
            if lista[i]=='1' and flag=='1':
                lista[i]='0'
                flag='1'
            elif lista[i]=='0' and flag=='1':
                lista[i]='1'
                flag='0'
            elif lista[i]=='0' and flag=='0':
                lista[i]='0'
                flag='0'
            i+=1
        
        if flag=='1':
            lista.append('1')
        rep=''
        for s in lista[::-1]:
            rep=rep+s
        return rep
        
###################################################################
class Solution:
    def addBinary(self, a, b):
        """
        :type a: str
        :type b: str
        :rtype: str
        """
        return bin(int(a,2)+int(b,2))[2:]

File: archives/C#/0067. Add Binary.cs
public class Solution {
    public string AddBinary(string a, string b) {
        if(a.Length<b.Length){
            string tmp=a;
            a=b;
            b=tmp;
        }
        int flag='0';
        string rep="";
        for(int i=a.Length-1;i>=0;i--){
            if(i>=a.Length-b.Length){
                if(a[i]=='0' && b[i+b.Length-a.Length]=='0' && flag=='0'){
                    rep='0'+rep;
                }
                else if(a[i]=='1' && b[i+b.Length-a.Length]=='0' && flag=='0'){
                    rep='1'+rep;
                }
                else if(a[i]=='0' && b[i+b.Length-a.Length]=='1' && flag=='0'){
                    rep='1'+rep;
                }
                else if(a[i]=='1' && b[i+b.Length-a.Length]=='1' && flag=='0'){
                    rep='0'+rep;
                    flag='1';
                }else if(a[i]=='0' && b[i+b.Length-a.Length]=='0' && flag=='1'){
                    rep='1'+rep;
                    flag='0';
                }
                else if(a[i]=='1' && b[i+b.Length-a.Length]=='0' && flag=='1'){
                    rep='0'+rep;
                    flag='1';
                }
                else if(a[i]=='0' && b[i+b.Length-a.Length]=='1' && flag=='1'){
                    rep='0'+rep;
                    flag='1';
                }
                else if(a[i]=='1' && b[i+b.Length-a.Length]=='1' && flag=='1'){
                    rep='1'+rep;
                    flag='1';
                }
                
            }
            else{
                if(a[i]=='0' && flag=='0'){
                    rep='0'+rep;
                }
                else if(a[i]=='1' && flag=='0'){
                    rep='1'+rep;
                }
                else if(a[i]=='1' && flag=='1'){
                    rep='0'+rep;
                }
                else{
                    rep='1'+rep;
                    flag='0';
                }
            }
        }
        if (flag=='1'){
            return '1'+rep;
        }
        else{
            return rep;
        }
    }
}
